Order wrapper class proposal

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR is a proposal for introducing a new children class (`WC_Stripe_Order`, extended from `WC_Order`) that wraps all our custom metadata and extension-specific business logic. This should make it easier to reference and change metas.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Take a look at the code. Do you think it makes sense to move forward?
- Check if the tests are still passing
- Perform some smoke tests:
-- Connect your Stripe account
-- Enable some payment methods
-- As a shopper, make some purchases

\* Unit tests for the new class will be added if this proposal gets approved

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Order wrapper class proposal

wjrosa committed:
Including more fields

wjrosa commented:
Hey team! I would like to have your opinion on this proposal (@Mayisha , @diegocurbelo , @mattallan , @annemirasol ). What do you think?

annemirasol commented:
I love that we're consolidating order metadata getters and setters, and abstracting specifics, like the actual key names.

I haven't looked at all the changes closely, but I think it's a refactor worth pursuing, @wjrosa. Thanks for making the effort.

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa commented:
> I love that we're consolidating order metadata getters and setters, and abstracting specifics, like the actual key names.
> 
> I haven't looked at all the changes closely, but I think it's a refactor worth pursuing, @wjrosa. Thanks for making the effort.

Nice, Anne! I will update this PR and properly move this to review so everyone can have a better look at the idea.

wjrosa committed:
Additional getters and setters

wjrosa committed:
Additional getters and setters

wjrosa committed:
Lock/unlock methods

wjrosa committed:
Currency, net and fee getters and setters

wjrosa committed:
Min amount validator

wjrosa committed:
Get owner details

wjrosa committed:
Fix missing include

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Replacing get_order calls

wjrosa committed:
Fix tests

wjrosa committed:
Moving order methods to the new order class

wjrosa committed:
Moving meta query methods

wjrosa committed:
Abstracting meta query methods

wjrosa commented:
Hey @woocommerce/quark! I have finished polishing my idea here. I'd appreciate it if you could at least tell me if it is OK to move this forward. I believe it is a great quality of life improvement. All possible meta keys, getters, setters, and retrieval methods are in one place. This also makes our WC_Stripe_Helper class smaller.

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

Mayisha reviewed (commented):
> This PR is a proposal for introducing a new children class (WC_Stripe_Order, extended from WC_Order) that wraps all our custom metadata and extension-specific business logic.

This looks like a good refactor to me. 👍 

I am not done smoke testing all the flows. But here are a few points @wjrosa -
- There are a few conflicts with the latest changes. Let's do the functional tests after they are resolved.
- We need to move [lock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1722) and [unlock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1750C18-L1750C37) to `WC_Stripe_Order` class. Please recheck the latest `develop` branch to identify similar missing functions/usages that were merged after this PR was created.
- I have tested a simple product purchase and got this error in the logs-

<img width="1291" alt="Screenshot 2025-01-07 at 8 05 52 PM" src="https://github.com/user-attachments/assets/f25eec1e-19d3-4ead-b215-06e26a79487e" />

diegocurbelo commented on includes/class-wc-stripe-order.php:
```diff
@@ -0,0 +1,740 @@
+	 *
+	 * @return bool Whether the payment is awaiting action.
+	 */
+	public function payment_awaiting_action() {
```
For consistency, should be `is_payment_awaiting_action` or `get_payment_awaiting_action `

wjrosa replied:
Addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/1dc95df6d048882b482e40171c6af42b7ce32796

diegocurbelo commented on includes/class-wc-stripe-order.php:
```diff
@@ -0,0 +1,740 @@
+	 *
+	 * @return bool The value of the flag.
+	 */
+	public function upe_redirect_processed() {
```
For consistency, should be `is_ upe_redirect_processed ` or `get_ upe_redirect_processed`

wjrosa replied:
Addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/1dc95df6d048882b482e40171c6af42b7ce32796

diegocurbelo commented on includes/class-wc-stripe-order.php:
```diff
@@ -0,0 +1,740 @@
+	 *
+	 * @return bool
+	 */
+	public function upe_waiting_for_redirect() {
```
For consistency, this should be `is_upe_waiting_for_redirect` or `get_upe_waiting_for_redirect `

wjrosa replied:
Addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/1dc95df6d048882b482e40171c6af42b7ce32796

diegocurbelo commented on includes/class-wc-stripe-order.php:
```diff
@@ -0,0 +1,740 @@
+	 *
+	 * @return bool
+	 */
+	public function status_final() {
```
For consistency, this should be `is_status_final` or `get_status_final`

wjrosa replied:
Addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/1dc95df6d048882b482e40171c6af42b7ce32796

diegocurbelo reviewed (changes requested):
This looks awesome @wjrosa 👏🏼, I like that it makes the code easier to understand, and we consolidate the string constants for the meta table in one place... 

I added a few function re-name suggestions for consistency... we normally use `is_` or `has_` as preffix for functions that return bool values.

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Renaming Stripe Order methods for consistency

wjrosa commented:
> > This PR is a proposal for introducing a new children class (WC_Stripe_Order, extended from WC_Order) that wraps all our custom metadata and extension-specific business logic.
> 
> This looks like a good refactor to me. 👍
> 
> I am not done smoke testing all the flows. But here are a few points @wjrosa -
> 
> * There are a few conflicts with the latest changes. Let's do the functional tests after they are resolved.
> * We need to move [lock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1722) and [unlock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1750C18-L1750C37) to `WC_Stripe_Order` class. Please recheck the latest `develop` branch to identify similar missing functions/usages that were merged after this PR was created.
> * I have tested a simple product purchase and got this error in the logs-
> 
> <img alt="Screenshot 2025-01-07 at 8 05 52 PM" width="1291" src="https://private-user-images.githubusercontent.com/33387139/400817539-f25eec1e-19d3-4ead-b215-06e26a79487e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NzQ5NTIsIm5iZiI6MTczODg3NDY1MiwicGF0aCI6Ii8zMzM4NzEzOS80MDA4MTc1MzktZjI1ZWVjMWUtMTlkMy00ZWFkLWIyMTUtMDZlMjZhNzk0ODdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDIwNDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxYTgwYmUwMjhhMDExODhhNzE1ZDdiMzBiNDlhNDEwZDkyYjU0ODhmNzZlMjQ3OGIzNmY3NjA2NGZjOWUxMzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8OX_hoC4PoowVvwBQLqcKZHHKnUinKNxfDGgy20c5AM">

Thanks, @Mayisha ! 

> * There are a few conflicts with the latest changes. Let's do the functional tests after they are resolved.

Solved ✅ 

> * We need to move [lock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1722) and [unlock_order_refund](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1750C18-L1750C37) to `WC_Stripe_Order` class. Please recheck the latest `develop` branch to identify similar missing functions/usages that were merged after this PR was created.

Done ✅ 

> * I have tested a simple product purchase and got this error in the logs-

I could not reproduce this. Is this still happening for you? It is strange because the error implies that we are passing the `Automattic\WooCommerce\Admin\Overrides\Order` class to `get_intent_from_order`, which would make sense only if we did inside the extension. But I could not find any method instantiating this class on our code. What I can do is add a check for the class (on it). Edit: done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/c7fc075e5d5b1155eb9dfab0d5b47b119671f547

wjrosa committed:
Adding fallback to avoid fatal error

wjrosa committed:
Renaming additional methods + unit tests

diegocurbelo committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

diegocurbelo committed:
Merge branch 'develop' into dev/order-wrapper-proposal

diegocurbelo reviewed (changes requested):
Thanks for all the changes @wjrosa!..

I found a regression though, the full refund E2E test is failing:

![Screenshot_2025-02-12_at_17_08_28](https://github.com/user-attachments/assets/f96599ae-d167-4106-9b77-bb456ec75588)


This is the error log:
```
[12-Feb-2025 19:52:33 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method Automattic\WooCommerce\Admin\Overrides\Order::get_upe_payment_type() in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:286
Stack trace:
#0 /var/www/html/wp-content/plugins/woocommerce/includes/admin/meta-boxes/views/html-order-items.php(384): WC_Stripe_UPE_Payment_Gateway->can_refund_order(Object(Automattic\WooCommerce\Admin\Overrides\Order))
#1 /var/www/html/wp-content/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-items.php(40): include('/var/www/html/w...')
#2 /var/www/html/wp-admin/includes/template.php(1456): WC_Meta_Box_Order_Items::output(NULL, Array)
#3 /var/www/html/wp-content/plugins/woocommerce/src/Internal/Admin/Orders/Edit.php(501): do_meta_boxes(Object(WP_Screen), 'normal', Object(Automattic\WooCommerce\Admin\Overrides\Order))
#4 /var/www/html/wp-content/plugins/woocommerce/src/Internal/Admin/Orders/Edit.php(405): Automattic\WooCommerce\Internal\Admin\Orde in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php on line 286
```

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Fix regression

wjrosa commented:
> Thanks for all the changes @wjrosa!..
> 
> I found a regression though, the full refund E2E test is failing:
> 
> ![Screenshot_2025-02-12_at_17_08_28](https://private-user-images.githubusercontent.com/407542/412597100-f96599ae-d167-4106-9b77-bb456ec75588.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NDg4NDcsIm5iZiI6MTczOTQ0ODU0NywicGF0aCI6Ii80MDc1NDIvNDEyNTk3MTAwLWY5NjU5OWFlLWQxNjctNDEwNi05Yjc3LWJiNDU2ZWM3NTU4OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxM1QxMjA5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hOGFhMGMxZDIwNjY1MTBkZGY4MjJhNGQyYWZkNDhjM2FmZTk3NWY3NDM0MzQ3ZDA2OWUyZDE2ZDg2YjRjZjg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Jq7h4B0P6GoBSwesgtjqkumwSxssxAQ2w4bLNvtH24w)
> 
> This is the error log:
> 
> ```
> [12-Feb-2025 19:52:33 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method Automattic\WooCommerce\Admin\Overrides\Order::get_upe_payment_type() in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:286
> Stack trace:
> #0 /var/www/html/wp-content/plugins/woocommerce/includes/admin/meta-boxes/views/html-order-items.php(384): WC_Stripe_UPE_Payment_Gateway->can_refund_order(Object(Automattic\WooCommerce\Admin\Overrides\Order))
> #1 /var/www/html/wp-content/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-items.php(40): include('/var/www/html/w...')
> #2 /var/www/html/wp-admin/includes/template.php(1456): WC_Meta_Box_Order_Items::output(NULL, Array)
> #3 /var/www/html/wp-content/plugins/woocommerce/src/Internal/Admin/Orders/Edit.php(501): do_meta_boxes(Object(WP_Screen), 'normal', Object(Automattic\WooCommerce\Admin\Overrides\Order))
> #4 /var/www/html/wp-content/plugins/woocommerce/src/Internal/Admin/Orders/Edit.php(405): Automattic\WooCommerce\Internal\Admin\Orde in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php on line 286
> ```

Oops! Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/b43b30e9b5b5c247ec7932129b329f593446b844

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

diegocurbelo committed:
Merge branch 'develop' into dev/order-wrapper-proposal

Mayisha committed:
Merge branch 'develop' into dev/order-wrapper-proposal

Mayisha commented on includes/class-wc-stripe-helper.php:
```diff
@@ -9,12 +9,47 @@
+	 * @deprecated 9.1.0
+	 */
+	const LEGACY_META_NAME_FEE = 'Stripe Fee';
+	/**
```
These constants are unused now.

wjrosa replied:
On previous PRs, I was advised to deprecate constants to avoid breaking custom merchant code (using them). That's why I kept them. Do you think it is not needed?

Mayisha replied:
I did not think about this scenario. We should keep it then 👍

Mayisha commented on includes/class-wc-stripe-webhook-handler.php:
```diff
@@ -246,7 +246,7 @@ public function process_webhook_payment( $notification, $retry = true ) {
 		}
 
-		$order = WC_Stripe_Helper::get_order_by_source_id( $notification->data->object->id );
+		$order = WC_Stripe_Order::get_by_id( $notification->data->object->id );
```
```suggestion
		$order = WC_Stripe_Order::get_by_source_id( $notification->data->object->id );
```

Mayisha commented on includes/class-wc-stripe-webhook-handler.php:
```diff
@@ -615,11 +615,11 @@ public function process_webhook_charge_failed( $notification ) {
 		// If can't find order by charge ID, try source ID.
 		if ( ! $order ) {
-			$order = WC_Stripe_Helper::get_order_by_source_id( $notification->data->object->id );
+			$order = WC_Stripe_Order::get_by_id( $notification->data->object->id );
```
```suggestion
			$order = WC_Stripe_Order::get_by_source_id( $notification->data->object->id );
```

I think this is the correct function.

Mayisha reviewed (changes requested):
### Smoke test results
**Account connection**
✅ Connect to Stripe account
✅ Reconfigure webhook

**Simple product**
✅ Card
✅ WeChat
✅ Multibanco
✅ klarna
✅ iDeal
⚠️ Sepa - The status remained on-hold forever. (check if you can reproduce it or it could be my local issue)

**Subscription**
✅ Card
✅ Cash App
⚠️ Sepa - The status remained on-hold (check if you can reproduce it or it could be my local issue)
🔴 Renewal - subscription renewal from subs edit page results in a fatal error
<img width="840" alt="Screenshot 2025-02-25 at 4 07 31 PM" src="https://github.com/user-attachments/assets/d62e78ed-b562-4037-84f9-4a7e6d97099a" />

**Refund**
✅ Tested refund with a couple of payment methods for simple and subscription product.

**Capture Later**
✅ Tested with card.

### Additional note
**Changes needed**
- Need to update the deprecation version from `9.1.0` to the latest one.
- Added a few comments.

**Post merge action item**
- Create a follow-up issue to remove deprecated functions in the future.
- Inform Quark and Fractal members to use this wrapper going forward.
- Include rigorous testing instructions for order related features for GS before the release.

wjrosa committed:
Fix tests

wjrosa committed:
Update includes/class-wc-stripe-webhook-handler.php

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>

wjrosa committed:
Update includes/class-wc-stripe-webhook-handler.php

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>

wjrosa committed:
Update deprecation comments

wjrosa committed:
Fix fatal error

wjrosa committed:
Fix query by transaction (charge) id

wjrosa commented:
Thanks @Mayisha for the additional review and notes!

> ⚠️ Sepa - The status remained on-hold forever. (check if you can reproduce it or it could be my local issue)

Fixed this issue in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/8be7874a9c43bad6cf60a12b7e8f4bf2717bd5f1

> 🔴 Renewal - subscription renewal from subs edit page results in a fatal error

Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3654/commits/d64f1548c4eeddaf33c412af79063c85217881d9

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

Mayisha commented on includes/class-wc-stripe-helper.php:
```diff
@@ -61,144 +97,142 @@ public static function delete_main_stripe_settings() {
 
-		$order->delete_meta_data( self::META_NAME_FEE );
-		$order->delete_meta_data( self::LEGACY_META_NAME_FEE );
+		$order->delete_fee();
```
```suggestion
		$order->delete_meta_data( self::META_NAME_FEE );
		$order->delete_meta_data( self::LEGACY_META_NAME_FEE );
```


@wjrosa I am still getting the fatal error when I try to renew a subscription from the subscription edit page. The reason is , though it's calling this helper method now, the helper method itself tries to call the function from the new order wrapper class.

<img width="916" alt="Screenshot 2025-02-27 at 6 05 04 PM" src="https://github.com/user-attachments/assets/8c5159fa-1c5f-4b1b-b667-0aa3a6b3ba8d" />

This helps to realize another problem, if any third party plugins are using these helper methods, they would send an instance of `WC_Order` here, not `WC_Stripe_Order`. So we should keep the previous code in these functions. 

Similar changes (reverts) should be made for other functions here as well to  avoid breaking plugin compaatibility.

wjrosa replied:
Alright, sorry about that! I reverted the changes on the deprecated methods and added extra checks for the object class type on the latest commits.

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Making deprecated methods use original code

wjrosa committed:
Updating PHP docs

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Making deprecated methods use original code

wjrosa committed:
Compatibility fixes

wjrosa committed:
Compatibility fixes

wjrosa committed:
Removing unused method

wjrosa committed:
Adding constants for meta keys

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Fix renewal subscription issues

wjrosa committed:
Helper method to convert an WC_Order into an WC_Stripe_Order

wjrosa committed:
Specific unit test

wjrosa committed:
Fix tests + moving instance method up

Mayisha committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa committed:
Merge branch 'develop' into dev/order-wrapper-proposal

wjrosa commented:
Closing in favor of the separated PRs.
